crowbar_framework/app/models/crowbar_service.rb

```diff
@@ -32,6 +32,79 @@ def role_constraints
     end
   end
 
+  def shutdown_openstack_at_nodes
+    db_nodes = []
+    NodeObject.find("state:crowbar_upgrade").each do |node|
+
+      # Find the nodes with Database role
+      # (we can't look at node.roles, because that list has been reduced to just crowbar-upgrade)
+      node.crowbar["run_list_map"].each do |role, v|
+        db_nodes.push node.name if role == "database-server"
+      end
+
+      # mark the position in the upgrade process
+      node["crowbar_wall"]["crowbar_openstack_shutdown"] = true
+      node.save
+    end
+
+    # Adapt current proposal, so only non-db nodes have crowbar-upgrade role
+    # (crowbar-upgrade role is currently assigned to all nodes)
+    proposal = ProposalObject.find_proposal("crowbar", "default")
+    proposal["deployment"]["crowbar"]["elements"]["crowbar-upgrade"] -= db_nodes
+    proposal.save
+
+    return true if proposal["deployment"]["crowbar"]["elements"]["crowbar-upgrade"].empty?
+
+    # Commit proposal so the shutdown actions from crowbar-upgrade get executed for non-db nodes
+    proposal_commit("default", false, false)
+  end
+
+  def dump_database
+    db_nodes = []
+    NodeObject.find("state:crowbar_upgrade").each do |node|
+
+      # Find the nodes with Database role
+      # (we can't look at node.roles, because that list has been reduced to just crowbar-upgrade)
+      node.crowbar["run_list_map"].each do |role, v|
+        db_nodes.push node.name if role == "database-server"
+      end
+    end
+
+    proposal = ProposalObject.find_proposal("crowbar", "default")
+    # After all non-DB services are shut down, adapt the proposal again and
+    # commit it with DB nodes only (and the special role for them)
+    proposal["deployment"]["crowbar"]["elements"]["crowbar-upgrade"] = []
+    proposal["deployment"]["crowbar"]["elements"]["crowbar-db-dump"] = db_nodes
+    proposal.save
+
+    # This proposal could return some error if there's not enough space for DB dump
+    # Controller must show the error and be able to call the function again once the problem
+    # is resolved
+    proposal_commit("default", false, false)
+
+    proposal = ProposalObject.find_proposal("crowbar", "default")
+    if proposal["deployment"]["crowbar"]["crowbar-status"] == "failed"
+      raise proposal["deployment"]["crowbar"]["crowbar-failed"]
+    end
```

Comment:
It's because crowbar_service.rb is overriding `apply_role` method, and although it calls `super` in it, it ignores the result of the one called from superclass: 

https://github.com/crowbar/crowbar-core/blob/master/crowbar_framework/app/models/crowbar_service.rb#L219